src/components/LibrariesPage.tsx

```diff
@@ -83,6 +96,36 @@ export class LibrariesPage extends React.Component<LibrariesPageProps, Libraries
     return libraries.filter(l => l.stages.registry_stage === "production");
   }
 
+  setFilter(filter: string) {
+    let updatedFilters = this.state.filters.attributes;
+    updatedFilters[filter] = !this.state.filters.attributes[filter];
+    this.setState({ filters: {flipped: this.state.filters.flipped, attributes: updatedFilters} });
+  }
+
+  flipFilter() {
+    this.setState({ filters: {flipped: !this.state.filters.flipped, attributes: this.state.filters.attributes} });
+  }
+
+  filterByAttribute(libraries) {
+    let hasAttr = (dict, attr) => {
```

Comment:
This is a nice recursive function but there are no tests for this. Can you make this a separate function outside of `filterByAttribute` so you can test both `hasAttr` and `filterByAttribute`?